packages/runtime/container-runtime/src/pendingStateManager.ts

```diff
@@ -347,80 +335,52 @@ export class PendingStateManager implements IDisposable {
+             * Note: It is not possible for the PendingStateManager to receive a partially acked batch. It will
+             * either receive the whole batch ack or nothing at all.
+             */
+            if (pendingMessage.opMetadata?.batch) {
+                this.stateHandler.orderSequentially(() => {
+                    do {
                         this.stateHandler.reSubmit(
-                            pendingState.messageType,
-                            pendingState.content,
-                            pendingState.localOpMetadata,
-                            pendingState.opMetadata);
-                    }
-                    break;
-                case "flush":
-                    /**
-                     * A "flush" call can indicate the end of a batch.
-                     * We can't rely on the "batch" property in the message metadata as it gets
-                     * updated elsewhere and it is not the same object instance that gets updated.
-                     */
-                    if (messageBatchQueue.length > 0) {
-                        this.stateHandler.orderSequentially(() => {
-                            while (messageBatchQueue.length > 0) {
-                                // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-                                const message = messageBatchQueue.dequeue()!;
-                                this.stateHandler.reSubmit(
-                                    message.messageType,
-                                    message.content,
-                                    message.localOpMetadata,
-                                    message.opMetadata);
-                            }
-                        });
-                    }
-                    assert(messageBatchQueue.length === 0, 0x41c /* cannot flush in the middle of a batch */);
-                    this.stateHandler.flush();
-                    break;
-                default:
-                    break;
+                            pendingMessage.messageType,
+                            pendingMessage.content,
+                            pendingMessage.localOpMetadata,
+                            pendingMessage.opMetadata);
+
+                        if (pendingMessage.opMetadata?.batch === false) {
+                            break;
+                        }
+
+                        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+                        pendingMessage = this.pendingMessages.shift()!;
+                        pendingStatesCount--;
+                    } while (pendingStatesCount > 0);
+                });
+            } else {
+                this.stateHandler.reSubmit(
+                    pendingMessage.messageType,
+                    pendingMessage.content,
+                    pendingMessage.localOpMetadata,
+                    pendingMessage.opMetadata);
             }
             pendingStatesCount--;
```

Comment:
I might be misunderstanding this, but is should `pendingStatesCount--` be inside the else statement? Isn't this double decrementing in case of batch messages?

Reply:
The logic is technically correct as-is, but it is definitely confusing. I rearranged some of it to hopefully be clearer about checks.